### PR TITLE
Feature/issue 2539

### PR DESCRIPTION
**Pull request by kartikmodi**

### CODE REVIEW

File: docs/source/tutorial.ipynb [hunk 1]
```diff
@@ -303,7 +473,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.10"
+   "version": "3.6.15"
```
Comment (kartikmodi) [inline docs/source/tutorial.ipynb:476]: I have put down a 3.6.x python version as we support 3.6.x as what I learned from @andre-merzky 's comments in the chat history somewhere.

**Comment by andre-merzky** [comment]
I am missing something as I can't build the documentation just yet:
```sh
(ve3)  rivendell  merzky  …/merzky/radical/radical.pilot.work  [feature/issue_2539] $ sphinx-build -b html -d build/doctrees  docs/source/ build/html
Running Sphinx v1.6.7
loading translations [python]... not available for built-in messages

Extension error:
Could not import extension nbsphinx (exception: No module named 'nbsphinx')
(ve3)  rivendell  merzky  …/merzky/radical/radical.pilot.work  1  [feature/issue_2539] $ pip install -r docs/source/requirements.txt 
Requirement already satisfied: nbsphinx in ./ve3/lib/python3.8/site-packages (from -r docs/source/requirements.txt (line 1)) (0.8.8)
```
Do you have any advise?

**Comment by kartikmodi** [comment]
> Do you have any advise?
I see you are using a very old version of Sphinx. Maybe that could be a reason.
I am running 4.4.0.

Another guess is, you have installed Sphinx in virtualenv and not using the system's Sphinx from OS's package manager.

**Comment by kartikmodi** [comment]
@mturilli I have removed the output of the notebook from the notebook src which it much cleaner. The diff is now significantly shorter, and diff makes more sense/approachable to review.

**Comment by andre-merzky** [comment]
I just tried and your comment helped (no idea why the original install of sphinx did not pull the newer version).  But alas I stumble over another problem which I don't see how to resolve:

```sh
$ cat requirements-docs.txt 

sphinx>=4.4
sphinx_copybutton
sphinx_rtd_theme
nbsphinx
pandoc
ipython
python-dotenv

$ virtualenv -p python3.8 ve3
...

$ . ve3/bin/activate
(ve3) $ pip install .
...
(ve3) $ pip install -r requirements-docs.txt
...

(ve3) $ cd docs
(ve3) $ make -n html
sphinx-build -b html -d build/doctrees   source build/html
echo
echo "Build finished. The HTML pages are in build/html."

(ve3) $ which sphinx-build; sphinx-build --version
/mnt/home/merzky/radical/radical.pilot.test/ve3/bin/sphinx-build
sphinx-build 4.4.0

(ve3) $  make html
sphinx-build -b html -d build/doctrees   source build/html
Running Sphinx v4.4.0
loading translations [python]... not available for built-in messages
WARNING: html_static_path entry '_static' does not exist
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 8 source files that are out of date
updating environment: [new config] 8 added, 0 changed, 0 removed
/mnt/home/merzky/radical/radical.pilot.test/docs/source/configuration.ipynb:: WARNING: rST localisation for language "python" not found.                                                    
/mnt/home/merzky/radical/radical.pilot.test/docs/source/configuration.ipynb:2: WARNING: No smart quotes defined for language "python".
/mnt/home/merzky/radical/radical.pilot.test/docs/source/quickstart.ipynb:9: WARNING: No smart quotes defined for language "python".
/mnt/home/merzky/radical/radical.pilot.test/docs/source/resources.ipynb:9: WARNING: No smart quotes defined for language "python".

Notebook error:
CellExecutionError in tutorial.ipynb:
------------------
import os
import sys

import radical.pilot as rp
import radical.utils as ru

%load_ext dotenv
%dotenv
print("loaded2")
os.getenv("RADICAL_PILOT_DBURL")
os.getcwd()
------------------

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Input In [1], in <module>
      4 import radical.pilot as rp
      5 import radical.utils as ru
----> 7 get_ipython().run_line_magic('load_ext', 'dotenv')
      8 get_ipython().run_line_magic('dotenv', '')
      9 print("loaded2")

File /mnt/home/merzky/radical/radical.pilot.work/ve3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:2204, in InteractiveShell.run_line_magic(self, magic_name, line, _stack_depth)
   2202     kwargs['local_ns'] = self.get_local_scope(stack_depth)
   2203 with self.builtin_trap:
-> 2204     result = fn(*args, **kwargs)
   2205 return result

File /mnt/home/merzky/radical/radical.pilot.work/ve3/lib/python3.8/site-packages/IPython/core/magics/extension.py:33, in ExtensionMagics.load_ext(self, module_str)
     31 if not module_str:
     32     raise UsageError('Missing module name.')
---> 33 res = self.shell.extension_manager.load_extension(module_str)
     35 if res == 'already loaded':
     36     print("The %s extension is already loaded. To reload it, use:" % module_str)

File /mnt/home/merzky/radical/radical.pilot.work/ve3/lib/python3.8/site-packages/IPython/core/extensions.py:76, in ExtensionManager.load_extension(self, module_str)
     69 """Load an IPython extension by its module name.
     70 
     71 Returns the string "already loaded" if the extension is already loaded,
     72 "no load function" if the module doesn't have a load_ipython_extension
     73 function, or None if it succeeded.
     74 """
     75 try:
---> 76     return self._load_extension(module_str)
     77 except ModuleNotFoundError:
     78     if module_str in BUILTINS_EXTS:

File /mnt/home/merzky/radical/radical.pilot.work/ve3/lib/python3.8/site-packages/IPython/core/extensions.py:92, in ExtensionManager._load_extension(self, module_str)
     90 if module_str not in sys.modules:
     91     with prepended_to_syspath(self.ipython_extension_dir):
---> 92         mod = import_module(module_str)
     93         if mod.__file__.startswith(self.ipython_extension_dir):
     94             print(("Loading extensions from {dir} is deprecated. "
     95                    "We recommend managing extensions like any "
     96                    "other Python packages, in site-packages.").format(
     97                   dir=compress_user(self.ipython_extension_dir)))

File /mnt/home/merzky/radical/radical.pilot.work/ve3/lib/python3.8/importlib/__init__.py:127, in import_module(name, package)
    125             break
    126         level += 1
--> 127 return _bootstrap._gcd_import(name[level:], package, level)

File <frozen importlib._bootstrap>:1014, in _gcd_import(name, package, level)

File <frozen importlib._bootstrap>:991, in _find_and_load(name, import_)

File <frozen importlib._bootstrap>:973, in _find_and_load_unlocked(name, import_)

ModuleNotFoundError: No module named 'dotenv'
ModuleNotFoundError: No module named 'dotenv'

You can ignore this error by setting the following in conf.py:

    nbsphinx_allow_errors = True

Makefile:48: recipe for target 'html' failed
make: *** [html] Error 2
```

(Sorry for the long dump).
Any idea how to continue?

**Comment by kartikmodi** [comment]
This is surprising. I ran again on my system with the command 
`sphinx-build -b html -d build/doctrees   source build/html` because you are running inside from `docs` dir .
It runs properly on my machine, executing every cell. See output - 
```
(radical-pilot-env) workstation@debian-x:~/workspace/PycharmProjects/radical.pilot/docs$ sphinx-build -b html -d build/doctrees   source build/html
Running Sphinx v4.4.0
loading translations [python]... not available for built-in messages
making output directory... done
WARNING: html_static_path entry '_static' does not exist
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 8 source files that are out of date
updating environment: [new config] 8 added, 0 changed, 0 removed
reading sources... [100%] tutorial1                                                                                                                                                                            
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/configuration.ipynb:: WARNING: rST localisation for language "python" not found.
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/configuration.ipynb:2: WARNING: No smart quotes defined for language "python".
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/quickstart.ipynb:9: WARNING: No smart quotes defined for language "python".
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/resources.ipynb:9: WARNING: No smart quotes defined for language "python".
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/tutorial.ipynb:9: WARNING: No smart quotes defined for language "python".
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/configuration.ipynb: WARNING: document isn't included in any toctree
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/resources.ipynb: WARNING: document isn't included in any toctree
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/tutorial1.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
writing output... [100%] tutorial1                                                                                                                                                                             
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/overview.rst:13: WARNING: undefined label: chapter_installation
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/overview.rst:18: WARNING: undefined label: chapter_installation
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/resources.ipynb:: WARNING: Could not lex literal_block as "ipython3". Highlighting skipped.
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
generating indices... genindex /home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
done
copying notebooks ... [100%] tutorial.ipynb                                                                                                                                                                    
writing additional pages... search /home/workstation/workspace/PycharmProjects/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
opensearch done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 23 warnings.

The HTML pages are in build/html.
```

**Another thing (which you already did, Sorry I missed) -** 
I was highly confused on how to pass env variables in notebooks i.e. either through the file say (.env) or setting env variable with %env (notebook magic) per notebook level. 
Because of this confusion, I avoided committing the `python-dotenv` dependency in requirements.txt, and later I forgot to properly deal with it. 
This is not approved/discussed because the priority is to write notebook documentation. 
We'll standardize/fix the env variable in what Matteo called `Level 1`, currently the focus is to roll out NB documentation `Level 0`.

I have committed it for now but marked it as #TODO in the documentation. 

Legends:
1. `Level 0` - NB working on local i.e. target: draft-v1
2. `Level 1` - NB documentation working on ReadTheDocs, within it's containers, with all validations (choices like external MongoDB connection, or embedded MongoDB usage, and env variable passing thing) 
I have updated the docs/source/README.md. 


**Guess Work for your issue, though highly unlikely to be useful this time** 
1. Ensuring dot-env is installed in virtual env's pip and system's  pip - I prefer to run command `python -m pip install python-dotenv` to get the right pip
2. Ensuring I don't have a system package manager dotenv package (like python3-dotenv / dotenv in case of apt)
3. Ensuring you have a `.env` file in {PROJECT_ROOT} dir. Though error would have been different. :smile: 
4. Install once again with python-dotenv[cli] for no reason, no logic here. [cli] just gives abilities to edit `.env` file without opening it editor.

**Review by mturilli — COMMENTED** [review]
I tested the PR on three.radical-project.org. It worked but I would suggest to better document the use or `.env`. Specifically, I would:

- Explain where the `.env` file needs to be located
- Use a general string for indicating the mongodb value, e.g., `mongodb://<user_id>:<password>@<ip_address>:<port>/<db_name>`

I would also look into these warnings:
```
writing output... [100%] tutorial1

/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/overview.rst:13: WARNING: undefined label: chapter_installation
/home/mturilli/github/radical.pilot/docs/source/overview.rst:18: WARNING: undefined label: chapter_installation
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/resources.ipynb:: WARNING: Could not lex literal_block as "ipython3". Highlighting skipped.
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
/home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
generating indices... genindex /home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
done

copying notebooks ... [100%] tutorial.ipynb

writing additional pages... search /home/mturilli/github/radical.pilot/docs/source/index.rst:21: WARNING: toctree contains reference to document 'developers' that doesn't have a title: no link will be generated
```

**Comment by mturilli** [comment]
As discussed, add also instructions about using requirements.txt to install the dependencies before running sphinx.

**Comment by mturilli** [comment]
As a side note, what is the benefit of using .env (and its module) over just exporting the variables?

**Comment by kartikmodi** [comment]
@mturilli 
> Explain where the .env file needs to be located 

I explained in README.md. Explained more so it's clearer. 

> Use a general string for indicating the mongodb value, e.g., mongodb://<user_id>:<password>@<ip_address>:<port>/<db_name>

From my experience, I have seen all configuration files defaults to localhost by convention, so I used localhost now.
I have included the `mongodb://<user_id>:<password>@<ip_address>:<port>/<db_name>` as format. See README.md now 

> instructions about using requirements.txt to install the dependencies before running sphinx.

I Missed it. DONE now.

> As a side note, what is the benefit of using .env (and its module) over just exporting the variables?

`.env` is coming originally from nodejs ecosystem (I think). i agree to the fact it's not native and still not widely accepted in python world. I am not contending to support `.env` in RCT. 
**The problem is with notebooks**. Notebooks have isolation for env variables, so for every notebook run, we have to specify the env variables **within** the notebooks. This is severe redundancy and a stupid idea to use NB's magic  `%env` since it's **not centralized.**

**To answer the question -** 
1. .env gives centralized control, exporting variables in every notebook will make things extremely hard, when someone wants to change a value of an env variable.

Other basic points - 
2. .env is a file so it gives versioning, history control. It makes things extremely easy to debug and audit. 
3. Since it's a file we can make automation on it more easily. (DevOps point of view)

**Review by mturilli — APPROVED** [review]
Thanks, the improvement look good and we can merge to enable level 0 of the documentation. We will then iterate and move towards level 1 with future PRs.

**Pull request merged**